File: sponge-kotlin/src/test/kotlin/examples/PluginsKb.kt
/*
 * Sponge Knowledge base
 * Defining plugins in a Kotlin knowledge base.
 */

package org.openksavi.sponge.kotlin.examples

import org.openksavi.sponge.config.Configuration
import org.openksavi.sponge.event.Event
import org.openksavi.sponge.examples.ConnectionPlugin
import org.openksavi.sponge.examples.EchoPlugin
import org.openksavi.sponge.kotlin.KKnowledgeBase
import org.openksavi.sponge.kotlin.KPlugin
import org.openksavi.sponge.kotlin.KTrigger

class PluginsKb : KKnowledgeBase() {

    override fun onInit() {
        // Variables for assertions only
        sponge.setVariable("valueBefore", null)
        sponge.setVariable("valueAfter", null)
    }

    /** Example plugin defined in the Kotlin knowledge base. */
    class KotlinKbPlugin : KPlugin() {
        var storedValue: Any? = null

        override fun onConfigure(configuration: Configuration) {
            storedValue = configuration.getString("storedValue", "default")
        }

        override fun onInit() = logger.debug("Initializing {}", name)

        override fun onStartup() = logger.debug("Starting up {}", name)
    }

    class PluginTrigger : KTrigger() {
        override fun onConfigure() {
            withEvent("e1")
        }
        override fun onRun(event: Event) {
            val kotlinKbPlugin = sponge.getPlugin(KotlinKbPlugin::class.java)
            val valueBefore = kotlinKbPlugin.storedValue
            logger.info("Plugin stored value: {}", valueBefore)
            sponge.setVariable("valueBefore", valueBefore)
            kotlinKbPlugin.storedValue = event.get<Any>("value")
            val valueAfter = kotlinKbPlugin.storedValue
            logger.info("New stored value: {}", valueAfter)
            sponge.setVariable("valueAfter", valueAfter)
        }
    }

    override fun onStartup() {
        sponge.event("e1").set("value", "Value B").send()
    }
}

File: sponge-jruby/examples/script/rb/plugins_kb.rb
# Sponge Knowledge Base
# Defining plugins in knowledge base.

def onInit
    # Variables for assertions only
    $sponge.setVariable("valueBefore", nil)
    $sponge.setVariable("valueAfter", nil)
end

# Example plugin defined in the knowledge base.
class ScriptPlugin < Plugin
    def onConfigure(configuration)
        @storedValue = configuration.getString("storedValue", "default")
    end
    def init
        self.logger.debug("Initializing {}", self.name)
    end
    def onStartup
        self.logger.debug("Starting up {}", self.name)
    end
    def getStoredValue
        return @storedValue
    end
    def setStoredValue(value)
        @storedValue = value
    end
end


class PluginTrigger < Trigger
    def onConfigure
        self.withEvent("e1")
    end
    def onRun(event)
        valueBefore = $scriptPlugin.getStoredValue()
        self.logger.info("Plugin stored value: {}", valueBefore)
        $sponge.setVariable("valueBefore", valueBefore)

        $scriptPlugin.setStoredValue(event.get("value"))
        valueAfter = $scriptPlugin.getStoredValue()
        self.logger.info("New stored value: {}", valueAfter)
        $sponge.setVariable("valueAfter", valueAfter)
    end
end

def onStartup
    $sponge.event("e1").set("value", "Value B").send()
end
